File: 06 June/06-10/itineraryArrangements.js
function getItineraryCount(stops) {
  const allStops = [...stops, "lunch", "dinner"];
  let validArrangements = 0;

  const backtrack = (pathSoFar, remainingStops) => {
    const breakfastIndex = pathSoFar.indexOf("breakfast");
    const lunchIndex = pathSoFar.indexOf("lunch");
    const dinnerIndex = pathSoFar.indexOf("dinner");
    let isValidArrangement = true;

    if (lunchIndex !== -1 && lunchIndex - breakfastIndex < 2
      || lunchIndex !== -1 && dinnerIndex !== -1 && dinnerIndex - lunchIndex < 2
      || lunchIndex === -1 && dinnerIndex !== -1
      || dinnerIndex !== -1 && pathSoFar.length - dinnerIndex > 2
    ) {
      isValidArrangement = false;
    }

    if (isValidArrangement) {
      validArrangements += pathSoFar.length === allStops.length + 1 ? 1 : 0;

      for (let i = 0; i < remainingStops.length; ++i) {
        pathSoFar.push(remainingStops[i]);
        backtrack(pathSoFar, remainingStops.toSpliced(i, 1));
        pathSoFar.pop();
      }
    }
  }

  backtrack(["breakfast"], allStops);

  return validArrangements;
}

// --- TEST SUITE ---

const testsText = String.raw`
1. getItineraryCount(["library", "park"]) should return 2.
2. getItineraryCount(["library", "park", "arcade"]) should return 18.
3. getItineraryCount(["library", "park", "arcade", "store"]) should return 120.
4. getItineraryCount(["library", "park", "arcade", "store", "cafe"]) should return 840.
5. getItineraryCount(["library", "park", "arcade", "store", "cafe", "market", "museum"]) should return 55440.
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("——————————————————————————",
            "\n🧪Starting Verification...",
            "\n——————————————————————————");

  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(`(${test.output})`);

    if (JSON.stringify(functionCallOutput) === JSON.stringify(testOutput)) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    } else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${JSON.stringify(testOutput)}\nGot: ${JSON.stringify(functionCallOutput)}`);

      ++failCount;
    }
    console.log("————————————————————————————");
  })

  console.log(failCount
    ? `⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`
    : "🎉SUCCESS: All tests PASSED."
  );
}

runTests(testData);

File: 06 June/06-10/itinerary_arrangements.py
def get_itinerary_count(stops):
    all_stops = stops + ["lunch", "dinner"]
    valid_arrangements = 0

    def backtrack(path_so_far, remaining_stops):
        nonlocal valid_arrangements
        breakfast_index = path_so_far.index("breakfast")

        try:
            lunch_index = path_so_far.index("lunch")
        except ValueError:
            lunch_index = None
        
        try:
            dinner_index = path_so_far.index("dinner")
        except ValueError:
            dinner_index = None

        is_valid_arrangement = True

        if lunch_index and lunch_index - breakfast_index < 2 \
        or lunch_index and dinner_index and dinner_index - lunch_index < 2 \
        or not lunch_index and dinner_index \
        or dinner_index and len(path_so_far) - dinner_index > 2:
            is_valid_arrangement = False

        if is_valid_arrangement:
            valid_arrangements += 1 if len(path_so_far) == len(all_stops) + 1 else 0

        for i, current_stop in enumerate(remaining_stops):
            path_so_far.append(current_stop)
            backtrack(path_so_far, [stop for stop in remaining_stops if stop != current_stop])
            path_so_far.pop()

    backtrack(["breakfast"], all_stops)

    return valid_arrangements

# --- TEST SUITE ---

tests_text = r'''
1. get_itinerary_count(["library", "park"]) should return 2.
2. get_itinerary_count(["library", "park", "arcade"]) should return 18.
3. get_itinerary_count(["library", "park", "arcade", "store"]) should return 120.
4. get_itinerary_count(["library", "park", "arcade", "store", "cafe"]) should return 840.
5. get_itinerary_count(["library", "park", "arcade", "store", "cafe", "market", "museum"]) should return 55440.
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("——————————————————————————",
        "\n🧪 Starting Verification...",
        "\n——————————————————————————")

    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            fail_count += 1

        print("————————————————————————————")

    print(
        f"⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED." if fail_count
        else "🎉 SUCCESS: All tests PASSED."
    )

run_tests(test_data)